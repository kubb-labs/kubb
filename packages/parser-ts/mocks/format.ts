import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string): Promise<string> {
  if (!source) {
    return Promise.resolve('')
  }
  return prettierFormat(source, formatOptions)
}
