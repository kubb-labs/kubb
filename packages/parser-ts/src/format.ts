import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}

/**
 * Format the generated code based on Prettier
 */
export async function format(source: string) {
  // do some basic linting with the ts compiler
  if (!source) {
    return ''
  }

  return prettierFormat(source, formatOptions)
}
