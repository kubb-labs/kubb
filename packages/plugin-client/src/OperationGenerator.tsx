import { createReactParser, OperationGenerator as Generator } from '@kubb/plugin-oas'
import { Oas } from '@kubb/plugin-oas/components'
import { App, createRoot } from '@kubb/react'

import { Client, Operations } from './components/index.ts'

import type { Operation } from '@kubb/oas'
import type { OperationMethodResult, OperationsByMethod } from '@kubb/plugin-oas'
import type { FileMeta, PluginClient } from './types.ts'

export const clientParser = createReactParser<PluginClient>({
  name: 'plugin-client',
  Operations({ options }) {
    if (!options.templates.operations) {
      return null
    }

    return <Operations.File baseURL={options.baseURL} templates={options.templates.operations} />
  },
  Operation({ options, operation }) {
    const isEnabled = options.client.methods.some((method) => operation.method === method)

    if (!options.templates.client || !isEnabled) {
      return null
    }

    return <Client.File baseURL={options.baseURL} templates={options.templates.client} />
  },
})
