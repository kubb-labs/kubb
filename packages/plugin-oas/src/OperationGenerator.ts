import { BaseGenerator, type FileMetaBase } from '@kubb/core'
import transformers from '@kubb/core/transformers'

import type { PluginFactoryOptions, PluginManager } from '@kubb/core'
import type * as KubbFile from '@kubb/fs/types'

import type { Plugin } from '@kubb/core'
import type { HttpMethod, Oas, OasTypes, Operation, contentType } from '@kubb/oas'
import type { Generator } from './generator.tsx'
import type { Exclude, Include, OperationSchemas, OperationsByMethod, Override } from './types.ts'

/**
 * @deprecated
 */
export type GetOperationGeneratorOptions<T extends OperationGenerator<any, any, any>> = T extends OperationGenerator<infer Options, any, any> ? Options : never

export type OperationMethodResult<TFileMeta extends FileMetaBase> = Promise<KubbFile.File<TFileMeta> | Array<KubbFile.File<TFileMeta>> | null>

type Context<TOptions, TPluginOptions extends PluginFactoryOptions> = {
  oas: Oas
  exclude: Array<Exclude> | undefined
  include: Array<Include> | undefined
  override: Array<Override<TOptions>> | undefined
  contentType: contentType | undefined
  pluginManager: PluginManager
  /**
   * Current plugin
   */
  plugin: Plugin<TPluginOptions>
  mode: KubbFile.Mode
}

export class OperationGenerator<
  TOptions = unknown,
  TPluginOptions extends PluginFactoryOptions = PluginFactoryOptions,
  TFileMeta extends FileMetaBase = FileMetaBase,
> extends BaseGenerator<TOptions, Context<TOptions, TPluginOptions>> {
  #operationsByMethod: OperationsByMethod = {}
  get operationsByMethod(): OperationsByMethod {
    return this.#operationsByMethod
  }

  set operationsByMethod(paths: OperationsByMethod) {
    this.#operationsByMethod = paths
  }

  #getOptions(operation: Operation, method: HttpMethod): Partial<TOptions> {
    const { override = [] } = this.context

    return (
      override.find(({ pattern, type }) => {
        if (type === 'tag') {
          return !!operation.getTags()[0]?.name.match(pattern)
        }

        if (type === 'operationId') {
          return !!operation.getOperationId().match(pattern)
        }

        if (type === 'path') {
          return !!operation.path.match(pattern)
        }

        if (type === 'method') {
          return !!method.match(pattern)
        }

        return false
      })?.options || {}
    )
  }

  /**
   *
   * @deprecated
   */
  #isExcluded(operation: Operation, method: HttpMethod): boolean {
    const { exclude = [] } = this.context
    let matched = false

    exclude.forEach(({ pattern, type }) => {
      if (type === 'tag' && !matched) {
        matched = !!operation.getTags()[0]?.name.match(pattern)
      }

      if (type === 'operationId' && !matched) {
        matched = !!operation.getOperationId().match(pattern)
      }

      if (type === 'path' && !matched) {
        matched = !!operation.path.match(pattern)
      }

      if (type === 'method' && !matched) {
        matched = !!method.match(pattern)
      }
    })

    return matched
  }

  /**
   *
   * @deprecated
   */
  #isIncluded(operation: Operation, method: HttpMethod): boolean {
    const { include = [] } = this.context
    let matched = false

    include.forEach(({ pattern, type }) => {
      if (type === 'tag' && !matched) {
        matched = !!operation.getTags()[0]?.name.match(pattern)
      }

      if (type === 'operationId' && !matched) {
        matched = !!operation.getOperationId().match(pattern)
      }

      if (type === 'path' && !matched) {
        matched = !!operation.path.match(pattern)
      }

      if (type === 'method' && !matched) {
        matched = !!method.match(pattern)
      }
    })

    return matched
  }

  getSchemas(
    operation: Operation,
    {
      forStatusCode,
      resolveName = (name) => name,
    }: {
      forStatusCode?: string | number
      resolveName?: (name: string) => string
    } = {},
  ): OperationSchemas {
    const pathParamsSchema = this.context.oas.getParametersSchema(operation, 'path')
    const queryParamsSchema = this.context.oas.getParametersSchema(operation, 'query')
    const headerParamsSchema = this.context.oas.getParametersSchema(operation, 'header')
    const requestSchema = this.context.oas.getRequestSchema(operation)
    const responseStatusCode =
      forStatusCode || (operation.schema.responses && Object.keys(operation.schema.responses).find((key) => key.startsWith('2'))) || 200
    const responseSchema = this.context.oas.getResponseSchema(operation, responseStatusCode)
    const statusCodes = operation.getResponseStatusCodes().map((statusCode) => {
      let name = statusCode
      if (name === 'default') {
        name = 'error'
      }

      const schema = this.context.oas.getResponseSchema(operation, statusCode)

      return {
        name: resolveName(transformers.pascalCase(`${operation.getOperationId()} ${name}`)),
        description: (operation.getResponseByStatusCode(statusCode) as OasTypes.ResponseObject)?.description,
        schema,
        operation,
        operationName: transformers.pascalCase(`${operation.getOperationId()}`),
        statusCode: name === 'error' ? undefined : Number(statusCode),
        keys: schema?.properties ? Object.keys(schema.properties) : undefined,
      }
    })

    return {
      pathParams: pathParamsSchema
        ? {
            name: resolveName(transformers.pascalCase(`${operation.getOperationId()} PathParams`)),
            operation,
            operationName: transformers.pascalCase(`${operation.getOperationId()}`),
            schema: pathParamsSchema,
            keys: pathParamsSchema.properties ? Object.keys(pathParamsSchema.properties) : undefined,
          }
        : undefined,
      queryParams: queryParamsSchema
        ? {
            name: resolveName(transformers.pascalCase(`${operation.getOperationId()} QueryParams`)),
            operation,
            operationName: transformers.pascalCase(`${operation.getOperationId()}`),
            schema: queryParamsSchema,
            keys: queryParamsSchema.properties ? Object.keys(queryParamsSchema.properties) : [],
          }
        : undefined,
      headerParams: headerParamsSchema
        ? {
            name: resolveName(transformers.pascalCase(`${operation.getOperationId()} HeaderParams`)),
            operation,
            operationName: transformers.pascalCase(`${operation.getOperationId()}`),
            schema: headerParamsSchema,
            keys: headerParamsSchema.properties ? Object.keys(headerParamsSchema.properties) : undefined,
          }
        : undefined,
      request: requestSchema
        ? {
            name: resolveName(transformers.pascalCase(`${operation.getOperationId()} ${operation.method === 'get' ? 'queryRequest' : 'mutationRequest'}`)),
            description: (operation.schema.requestBody as OasTypes.RequestBodyObject)?.description,
            operation,
            operationName: transformers.pascalCase(`${operation.getOperationId()}`),
            schema: requestSchema,
            keys: requestSchema.properties ? Object.keys(requestSchema.properties) : undefined,
            keysToOmit: requestSchema.properties
              ? Object.keys(requestSchema.properties).filter((key) => {
                  const item = requestSchema.properties?.[key] as OasTypes.SchemaObject

                  return item?.readOnly
                })
              : undefined,
          }
        : undefined,
      response: {
        name: resolveName(transformers.pascalCase(`${operation.getOperationId()} ${operation.method === 'get' ? 'queryResponse' : 'mutationResponse'}`)),
        description: operation.getResponseAsJSONSchema(responseStatusCode)?.at(0)?.description,
        operation,
        operationName: transformers.pascalCase(`${operation.getOperationId()}`),
        schema: responseSchema,
        statusCode: Number(responseStatusCode),
        keys: responseSchema?.properties ? Object.keys(responseSchema.properties) : undefined,
        keysToOmit: responseSchema?.properties
          ? Object.keys(responseSchema.properties).filter((key) => {
              const item = responseSchema.properties?.[key] as OasTypes.SchemaObject
              return item?.writeOnly
            })
          : undefined,
      },
      errors: statusCodes.filter((item) => item.statusCode?.toString().startsWith('4') || item.statusCode?.toString().startsWith('5')),
      statusCodes,
    }
  }

  #methods = ['get', 'post', 'patch', 'put', 'delete']

  async build(...generators: Array<Generator<Extract<TOptions, PluginFactoryOptions>>>): Promise<Array<KubbFile.File<TFileMeta>>> {
    const { oas } = this.context

    const paths = oas.getPaths()
    this.operationsByMethod = Object.entries(paths).reduce((acc, [path, method]) => {
      const methods = Object.keys(method) as HttpMethod[]

      methods.forEach((method) => {
        const operation = oas.operation(path, method)
        if (operation && [this.#methods].some((methods) => method === operation.method)) {
          const isExcluded = this.#isExcluded(operation, method)
          const isIncluded = this.context.include ? this.#isIncluded(operation, method) : true

          if (isIncluded && !isExcluded) {
            if (!acc[path]) {
              acc[path] = {} as OperationsByMethod['get']
            }
            acc[path] = {
              ...acc[path],
              [method]: {
                operation,
                schemas: this.getSchemas(operation),
              },
            } as OperationsByMethod['get']
          }
        }
      })

      return acc
    }, {} as OperationsByMethod)

    const promises = Object.keys(this.operationsByMethod).reduce((acc, path) => {
      const methods = this.operationsByMethod[path] ? (Object.keys(this.operationsByMethod[path]!) as HttpMethod[]) : []

      methods.forEach((method) => {
        const { operation } = this.operationsByMethod[path]?.[method]!
        const options = this.#getOptions(operation, method)

        const methodToCall = this[method as keyof typeof this] as any

        if (typeof methodToCall === 'function') {
          const promiseMethod = methodToCall?.call(this, operation, {
            ...this.options,
            ...options,
          })

          if (promiseMethod) {
            acc.push(promiseMethod)
          }
        }

        const promiseOperation = this.operation.call(this, operation, {
          ...this.options,
          ...options,
        })

        if (promiseOperation) {
          acc.push(promiseOperation)
        }

        generators?.forEach((generator) => {
          const promise = generator.operation?.({
            instance: this,
            operation,
            options: {
              ...this.options,
              ...options,
            },
          } as any) as Promise<Array<KubbFile.File<TFileMeta>>>

          if (promise) {
            acc.push(promise)
          }
        })
      })

      return acc
    }, [] as OperationMethodResult<TFileMeta>[])

    const operations = Object.values(this.operationsByMethod).map((item) => Object.values(item).map((item) => item.operation))

    promises.push(this.all(operations.flat().filter(Boolean), this.operationsByMethod))

    generators?.forEach((generator) => {
      const promise = generator.operations?.({
        instance: this,
        operations: operations.flat().filter(Boolean),
        operationsByMethod: this.operationsByMethod,
        options: this.options,
      } as any) as Promise<Array<KubbFile.File<TFileMeta>>>

      if (promise) {
        promises.push(promise)
      }
    })

    const files = await Promise.all(promises)

    // using .flat because operationGenerator[method] can return a array of files or just one file
    return files.flat().filter(Boolean)
  }

  /**
   * Operation
   */
  async operation(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }

  /**
   * GET
   */
  async get(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }

  /**
   * POST
   */
  async post(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }
  /**
   * PATCH
   */
  async patch(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }

  /**
   * PUT
   */
  async put(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }

  /**
   * DELETE
   */
  async delete(operation: Operation, options: TOptions): OperationMethodResult<TFileMeta> {
    return []
  }

  /**
   * Combination of GET, POST, PATCH, PUT, DELETE
   */
  async all(operations: Operation[], paths: OperationsByMethod): OperationMethodResult<TFileMeta> {
    return []
  }
}
