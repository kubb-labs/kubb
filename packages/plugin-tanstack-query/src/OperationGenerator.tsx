import { OperationGenerator as Generator } from '@kubb/plugin-oas'
import { Oas } from '@kubb/plugin-oas/components'
import { App, createRoot } from '@kubb/react'

import { Mutation } from './components/Mutation.tsx'
import { Query } from './components/Query.tsx'
import { QueryKey } from './components/QueryKey.tsx'
import { QueryOptions } from './components/QueryOptions.tsx'

import type { Operation } from '@kubb/oas'
import type { OperationMethodResult } from '@kubb/plugin-oas'
import { QueryImports } from './components/QueryImports.tsx'
import type { FileMeta, PluginTanstackQuery } from './types.ts'

export class OperationGenerator extends Generator<PluginTanstackQuery, FileMeta> {
  async operation(operation: Operation, options: PluginTanstackQuery['resolvedOptions']): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin, mode } = this.context

    const root = createRoot({
      logger: pluginManager.logger,
    })

    const templates = {
      mutation: Mutation.templates,
      query: Query.templates,
      queryOptions: QueryOptions.templates,
      queryKey: QueryKey.templates,
      queryImports: QueryImports.templates,
      ...options.templates,
    }

    const isQuery = typeof options.query === 'boolean' ? true : options.query?.methods.some((method) => operation.method === method)
    const isMutate = !isQuery && options.mutate && options.mutate.methods.some((method) => operation.method === method)

    root.render(
      <App pluginManager={pluginManager} plugin={{ ...plugin, options }} mode={mode}>
        <Oas oas={oas} operations={[operation]} generator={this}>
          <Oas.Operation operation={operation}>
            {isMutate && templates?.mutation && <Mutation.File templates={templates.mutation} />}
            {isQuery && templates?.query && templates.queryKey && templates.queryOptions && templates.queryImports && (
              <Query.File
                templates={{
                  query: templates.query,
                  queryKey: templates.queryKey,
                  queryOptions: templates.queryOptions,
                  queryImports: templates.queryImports,
                }}
              />
            )}
          </Oas.Operation>
        </Oas>
      </App>,
    )

    return root.files
  }
}
