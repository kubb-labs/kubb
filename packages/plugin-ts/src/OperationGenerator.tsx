import { OperationGenerator as Generator } from '@kubb/plugin-oas'
import { Oas } from '@kubb/plugin-oas/components'
import { App, createRoot } from '@kubb/react'

import { OasType } from './components/OasType.tsx'
import { OperationSchema } from './components/OperationSchema.tsx'

import type { Operation } from '@kubb/oas'
import type { OperationMethodResult } from '@kubb/plugin-oas'
import type { FileMeta, PluginTs } from './types.ts'
import { pascalCase } from '@kubb/core/transformers'

export class OperationGenerator extends Generator<PluginTs['resolvedOptions'], PluginTs> {
  async all(operations: Operation[]): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin, mode } = this.context

    const root = createRoot({
      logger: pluginManager.logger,
    })

    const Component = () => (
      <App pluginManager={pluginManager} plugin={plugin} mode={mode}>
        <Oas oas={oas} operations={operations} generator={this}>
          {plugin.options.oasType && <OasType.File name="oas" typeName="Oas" />}
        </Oas>
      </App>
    )

    Component.displayName = pascalCase('operations')

    root.render(<Component />)

    return root.files
  }

  async operation(operation: Operation, options: PluginTs['resolvedOptions']): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin, mode } = this.context

    const root = createRoot({
      logger: pluginManager.logger,
    })

    const Component = () => (
      <App pluginManager={pluginManager} plugin={{ ...plugin, options }} mode={mode}>
        <Oas oas={oas} operations={[operation]} generator={this}>
          <Oas.Operation operation={operation}>
            <OperationSchema.File />
          </Oas.Operation>
        </Oas>
      </App>
    )

    Component.displayName = pascalCase(operation.getOperationId())

    root.render(<Component />)

    return root.files
  }
}
