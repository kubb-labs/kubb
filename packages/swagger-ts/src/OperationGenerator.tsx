import { createRoot } from '@kubb/react'
import { OperationGenerator as Generator } from '@kubb/swagger'
import { Oas } from '@kubb/swagger/components'

import { OasType } from './components/OasType.tsx'
import { OperationSchema } from './components/OperationSchema.tsx'

import type { AppContextProps } from '@kubb/react'
import type { OperationMethodResult } from '@kubb/swagger'
import type { Operation } from '@kubb/swagger/oas'
import type { FileMeta, PluginOptions } from './types.ts'

export class OperationGenerator extends Generator<PluginOptions['resolvedOptions'], PluginOptions> {
  async all(operations: Operation[]): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })

    root.render(
      <Oas oas={oas} operations={operations} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        {plugin.options.oasType && <OasType.File name="oas" typeName="Oas" />}
      </Oas>,
      { meta: { pluginManager, plugin } },
    )

    return root.files
  }

  async operation(operation: Operation, options: PluginOptions['resolvedOptions']): OperationMethodResult<FileMeta> {
    const { oas, pluginManager, plugin, mode } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })
    root.render(
      <Oas oas={oas} operations={[operation]} getOperationSchemas={(...props) => this.getSchemas(...props)}>
        <Oas.Operation operation={operation}>
          <OperationSchema.File mode={mode} />
        </Oas.Operation>
      </Oas>,
      { meta: { pluginManager, plugin: { ...plugin, options } } },
    )

    return root.files
  }

  async get(): OperationMethodResult<FileMeta> {
    return null
  }

  async post(): OperationMethodResult<FileMeta> {
    return null
  }

  async put(): OperationMethodResult<FileMeta> {
    return null
  }
  async patch(): OperationMethodResult<FileMeta> {
    return null
  }
  async delete(): OperationMethodResult<FileMeta> {
    return null
  }
}
