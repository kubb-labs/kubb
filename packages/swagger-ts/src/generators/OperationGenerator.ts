import { getRelativePath } from '@kubb/core/utils'
import { print } from '@kubb/parser'
import * as factory from '@kubb/parser/factory'
import { OperationGenerator as Generator, resolve } from '@kubb/swagger'
import { camelCase, pascalCase } from 'change-case'

import { TypeBuilder } from '../builders/index.ts'

import type { KubbFile } from '@kubb/core'
import type { FileResolver, Operation, OperationSchemas, Resolver } from '@kubb/swagger'
import type ts from 'typescript'
import type { FileMeta, Options as PluginOptions } from '../types.ts'

type Options = {
  usedEnumNames: Record<string, number>

  mode: KubbFile.Mode
  enumType: NonNullable<PluginOptions['enumType']>
  dateType: NonNullable<PluginOptions['dateType']>
  optionalType: NonNullable<PluginOptions['optionalType']>
}

export class OperationGenerator extends Generator<Options> {
  resolve(operation: Operation): Resolver {
    const { pluginManager, plugin } = this.context

    return resolve({
      operation,
      resolveName: pluginManager.resolveName,
      resolvePath: pluginManager.resolvePath,
      pluginKey: plugin?.key,
    })
  }

  async all(): Promise<KubbFile.File | null> {
    return null
  }

  #printCombinedSchema(name: string, operation: Operation, schemas: OperationSchemas): string {
    const properties: Record<string, ts.TypeNode> = {
      'response': factory.createTypeReferenceNode(
        factory.createIdentifier(schemas.response.name),
        undefined,
      ),
    }

    if (schemas.request) {
      properties['request'] = factory.createTypeReferenceNode(
        factory.createIdentifier(schemas.request.name),
        undefined,
      )
    }

    if (schemas.pathParams) {
      properties['pathParams'] = factory.createTypeReferenceNode(
        factory.createIdentifier(schemas.pathParams.name),
        undefined,
      )
    }

    if (schemas.queryParams) {
      properties['queryParams'] = factory.createTypeReferenceNode(
        factory.createIdentifier(schemas.queryParams.name),
        undefined,
      )
    }

    if (schemas.headerParams) {
      properties['headerParams'] = factory.createTypeReferenceNode(
        factory.createIdentifier(schemas.headerParams.name),
        undefined,
      )
    }

    if (schemas.errors) {
      properties['errors'] = factory.createUnionDeclaration({
        nodes: schemas.errors.map(error => {
          return factory.createTypeReferenceNode(
            factory.createIdentifier(error.name),
            undefined,
          )
        }),
      })!
    }

    const namespaceNode = factory.createNamespaceDeclaration({
      name: operation.method === 'get' ? `${name}Query` : `${name}Mutation`,
      statements: Object.keys(properties).map(key => {
        const type = properties[key]
        if (!type) {
          return undefined
        }
        return factory.createTypeAliasDeclaration({
          modifiers: [factory.modifiers.export],
          name: pascalCase(key),
          type,
        })
      }).filter(Boolean),
    })

    return print(namespaceNode)
  }

  async get(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    const { mode, enumType, dateType, optionalType, usedEnumNames } = options
    const { pluginManager, plugin } = this.context

    const type = this.resolve(operation)

    const fileResolver: FileResolver = (name) => {
      // Used when a react-query type(request, response, params) has an import of a global type
      const root = pluginManager.resolvePath({ baseName: type.baseName, pluginKey: plugin?.key, options: { tag: operation.getTags()[0]?.name } })
      // refs import, will always been created with the SwaggerTS plugin, our global type
      const resolvedTypeId = pluginManager.resolvePath({
        baseName: `${name}.ts`,
        pluginKey: plugin?.key,
      })

      return getRelativePath(root, resolvedTypeId)
    }

    const source = new TypeBuilder({
      usedEnumNames,
      fileResolver: mode === 'file' ? undefined : fileResolver,
      withJSDocs: true,
      resolveName: (params) => pluginManager.resolveName({ ...params, pluginKey: plugin?.key }),
      enumType,
      optionalType,
      dateType,
    })
      .add(schemas.pathParams)
      .add(schemas.queryParams)
      .add(schemas.headerParams)
      .add(schemas.response)
      .add(schemas.errors)
      .configure()
      .print()

    const combinedSchemaSource = this.#printCombinedSchema(type.name, operation, schemas)

    return {
      path: type.path,
      baseName: type.baseName,
      source: [source, combinedSchemaSource].join('\n'),
      meta: {
        pluginKey: plugin.key,
        tag: operation.getTags()[0]?.name,
      },
    }
  }

  async post(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    const { mode, enumType, dateType, optionalType, usedEnumNames } = options
    const { pluginManager, plugin } = this.context

    const type = this.resolve(operation)

    const fileResolver: FileResolver = (name) => {
      // Used when a react-query type(request, response, params) has an import of a global type
      const root = pluginManager.resolvePath({ baseName: type.baseName, pluginKey: plugin?.key, options: { tag: operation.getTags()[0]?.name } })
      // refs import, will always been created with the SwaggerTS plugin, our global type
      const resolvedTypeId = pluginManager.resolvePath({
        baseName: `${name}.ts`,
        pluginKey: plugin?.key,
      })

      return getRelativePath(root, resolvedTypeId)
    }

    const source = new TypeBuilder({
      usedEnumNames,
      fileResolver: mode === 'file' ? undefined : fileResolver,
      withJSDocs: true,
      resolveName: (params) => pluginManager.resolveName({ ...params, pluginKey: plugin?.key }),
      enumType,
      optionalType,
      dateType,
    })
      .add(schemas.pathParams)
      .add(schemas.queryParams)
      .add(schemas.headerParams)
      .add(schemas.request)
      .add(schemas.response)
      .add(schemas.errors)
      .configure()
      .print()

    const combinedSchemaSource = this.#printCombinedSchema(type.name, operation, schemas)

    return {
      path: type.path,
      baseName: type.baseName,
      source: [source, combinedSchemaSource].join('\n'),
      meta: {
        pluginKey: plugin.key,
        tag: operation.getTags()[0]?.name,
      },
    }
  }

  async put(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
  async patch(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
  async delete(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
}
