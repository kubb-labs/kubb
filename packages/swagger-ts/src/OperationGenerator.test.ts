import { OasManager } from '@kubb/swagger'

import { format } from '../mocks/format.ts'
import { OperationGenerator } from './OperationGenerator.tsx'

import type { PluginContext, PluginManager } from '@kubb/core'
import type { KubbPlugin } from '@kubb/core'
import type { GetOperationGeneratorOptions } from '@kubb/swagger'
import type { PluginOptions } from './types.ts'

describe('OperationGenerator', () => {
  const resolvePath = () => './pets.ts'
  const resolveName: PluginContext['resolveName'] = ({ name }) => name

  it('[GET] should generate code based on a operation and optionalType `questionToken`', async () => {
    const oas = await OasManager.parseFromConfig({
      root: './',
      output: { path: 'test', clean: true },
      input: { path: 'packages/swagger-ts/mocks/petStore.yaml' },
    })

    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      enumType: 'asConst',
      dateType: 'string',
      optionalType: 'questionToken',
      usedEnumNames: {},
      transformers: {},
    }

    const og = await new OperationGenerator(
      options,
      {
        oas,
        exclude: [],
        include: undefined,
        pluginManager: { resolvePath, resolveName } as unknown as PluginManager,
        plugin: {} as KubbPlugin<PluginOptions>,
        contentType: undefined,
        override: undefined,
      },
    )
    const operation = oas.operation('/pets', 'get')

    const files = await og.get(operation, og.getSchemas(operation), options)
    const get = Array.isArray(files) ? files.at(0) : files

    expect(await format(get?.source)).toMatchSnapshot()

    const operationShowById = oas.operation('/pets/{petId}', 'get')

    const getShowByIdFiles = await og.get(operationShowById, og.getSchemas(operationShowById), options)
    const getShowById = Array.isArray(getShowByIdFiles) ? getShowByIdFiles.at(0) : getShowByIdFiles

    expect(await format(getShowById?.source)).toMatchSnapshot()
  })

  it('[POST] should generate code based on a operation', async () => {
    const oas = await OasManager.parseFromConfig({
      root: './',
      output: { path: 'test', clean: true },
      input: { path: 'packages/swagger-ts/mocks/petStore.yaml' },
    })

    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      enumType: 'asConst',
      dateType: 'string',
      optionalType: 'questionToken',
      usedEnumNames: {},
      transformers: {},
    }

    const og = await new OperationGenerator(
      options,
      {
        oas,
        exclude: [],
        include: undefined,
        pluginManager: { resolvePath, resolveName } as unknown as PluginManager,
        plugin: {} as KubbPlugin<PluginOptions>,
        contentType: undefined,
        override: undefined,
      },
    )
    const operation = oas.operation('/pets', 'post')

    const files = await og.post(operation, og.getSchemas(operation), options)
    const post = Array.isArray(files) ? files.at(0) : files

    expect(await format(post?.source)).toMatchSnapshot()
  })
})
