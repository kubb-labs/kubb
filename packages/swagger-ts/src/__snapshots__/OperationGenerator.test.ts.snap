// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OperationGenerator > [GET] should generate code based on a operation and optionalType \`questionToken\` 1`] = `
"export type ListPetsQueryParams = {
  /**
   * @description How many items to return at one time (max 100)
   * @type string | undefined
   */
  limit?: string
}

/**
 * @description unexpected error
 */
export type ListPetsError = Error

/**
 * @description A paged array of pets
 */
export type ListPetsQueryResponse = Pets
export namespace listPetsQuery {
  export type Response = ListPetsQueryResponse
  export type QueryParams = ListPetsQueryParams
  export type Errors = ListPetsError
}
"
`;

exports[`OperationGenerator > [GET] should generate code based on a operation and optionalType \`questionToken\` 2`] = `
"export type ShowPetByIdPathParams = {
  /**
   * @description The id of the pet to retrieve
   * @type string
   */
  petId: string
  /**
   * @description The id of the pet to retrieve
   * @type string
   */
  testId: string
}

/**
 * @description unexpected error
 */
export type ShowPetByIdError = Error

/**
 * @description Expected response to a valid request
 */
export type ShowPetByIdQueryResponse = Pet
export namespace showPetByIdQuery {
  export type Response = ShowPetByIdQueryResponse
  export type PathParams = ShowPetByIdPathParams
  export type Errors = ShowPetByIdError
}
"
`;

exports[`OperationGenerator > [POST] should generate code based on a operation 1`] = `
"/**
 * @description Null response
 */
export type CreatePets201 = any | null

export type CreatePetsMutationRequest = {
  /**
   * @type string
   */
  name: string
  /**
   * @type string
   */
  tag: string
}

export type CreatePetsMutationResponse = any | null

/**
 * @description unexpected error
 */
export type CreatePetsError = Error
export namespace createPetsMutation {
  export type Response = CreatePetsMutationResponse
  export type Request = CreatePetsMutationRequest
  export type Errors = CreatePets201 | CreatePetsError
}
"
`;
